File: rijndael-csharp/rijndael-encrypt.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;

public class RijndaelEncrypt {

  public static void Main( string[] args ) {
    if ( args.Length != 4 ) {
      Console.Error.WriteLine("Usage: " + Environment.GetCommandLineArgs()[0] + " <ifname> <ofname> <key> <iv>");
      Environment.Exit(1);
    }

    string ifname    = args[0];
    string ofname    = args[1];
    string key_fname = args[2];
    string iv_fname  = args[3];

    if( !File.Exists(ifname) ) {
      Console.Error.WriteLine("ABORT: input file not found: " + ifname);
      Environment.Exit(2);
    }

    byte[] KEY = hexStringToByteArray(Encoding.ASCII.GetString(fileReadBytes(key_fname)).Trim());
    byte[] IV  = hexStringToByteArray(Encoding.ASCII.GetString(fileReadBytes(iv_fname)).Trim());

    byte[] cipherBytes = rijndael_encrypt(fileReadBytes(ifname), ofname, KEY, IV);

    using HMACSHA512 hmac = new HMACSHA512(KEY);
    byte[] signature = hmac.ComputeHash(cipherBytes);
    fileWrite(ofname + ".sig", signature);
    Console.WriteLine("[ C# | encrypt ] signature : written");

    Array.Clear(KEY, 0, KEY.Length);
  }

  public static byte[] hexStringToByteArray( string s ) {
    int len = s.Length;
    byte[] data = new byte[len / 2];
    for (int i = 0; i < len; i += 2)
      data[i / 2] = (byte)((Convert.ToInt32(s[i].ToString(), 16) << 4) + Convert.ToInt32(s[i + 1].ToString(), 16));
    return data;
  }

  private static void fileWrite( string fileName, byte[] message ) {
    using FileStream fs = new FileStream(fileName, FileMode.Create, FileAccess.Write, FileShare.None);
    fs.Write(message, 0, message.Length);
  }

  private static byte[] fileReadBytes( string fname ) {
    try { return File.ReadAllBytes(fname); }
    catch (Exception e) { Console.Error.WriteLine("ABORT: " + e.Message); Environment.Exit(1); return null; }
  }

  public static byte[] rijndael_encrypt( byte[] plainBytes, string ofname, byte[] KEY, byte[] IV ) {
    using RijndaelManaged rijndael = new RijndaelManaged();
    rijndael.Key       = KEY;
    rijndael.IV        = IV;
    rijndael.Mode      = CipherMode.CBC;
    rijndael.KeySize   = 256;
    rijndael.BlockSize = 128;

    Console.WriteLine("[ C# | encrypt ] algorithm : AES-" + rijndael.KeySize + "/CBC/PKCS7");

    using ICryptoTransform encryptor = rijndael.CreateEncryptor(KEY, IV);
    using MemoryStream ms = new MemoryStream();
    using (CryptoStream cs = new CryptoStream(ms, encryptor, CryptoStreamMode.Write)) {
      cs.Write(plainBytes, 0, plainBytes.Length);
      cs.FlushFinalBlock();
    }
    byte[] cipherBytes = ms.ToArray();
    Console.WriteLine("[ C# | encrypt ] input     : " + plainBytes.Length + " bytes  ->  padded : " + cipherBytes.Length + " bytes");
    fileWrite(ofname, cipherBytes);
    Console.WriteLine("[ C# | encrypt ] output    : " + ofname);
    return cipherBytes;
  }
}

File: rijndael-java/RijndaelEncrypt.java
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.security.spec.*;
import java.util.Arrays;
import javax.crypto.*;
import javax.crypto.spec.*;

public class RijndaelEncrypt {

    private static byte[] fileRead(String fname) throws IOException {
        return Files.readAllBytes(Paths.get(fname));
    }

    private static void fileWrite(String fname, byte[] data) throws IOException {
        Files.write(Paths.get(fname), data);
    }

    private static byte[] hexToBytes(String hex) {
        int len = hex.length();
        byte[] data = new byte[len / 2];
        for (int i = 0; i < len; i += 2) {
            data[i / 2] = (byte) ((Character.digit(hex.charAt(i), 16) << 4)
                    + Character.digit(hex.charAt(i + 1), 16));
        }
        return data;
    }

    private static byte[] computeHmac(byte[] message, byte[] key) throws Exception {
        Mac hmac = Mac.getInstance("HmacSHA512");
        hmac.init(new SecretKeySpec(key, "HmacSHA512"));
        return hmac.doFinal(message);
    }

    public static void main(String[] args) {
        if (args.length != 4) {
            System.err.println("Usage: java -jar RijndaelEncrypt.jar <ifname> <ofname> <if_key> <if_iv>");
            System.exit(1);
        }

        String ifname = args[0];
        String ofname = args[1];
        String if_key = args[2];
        String if_iv  = args[3];

        File f = new File(ifname);
        if (!f.exists() || f.isDirectory()) {
            System.err.println("cannot find file: " + ifname);
            System.exit(2);
        }

        byte[] KEY = null;
        try {
            KEY = hexToBytes(new String(fileRead(if_key), StandardCharsets.US_ASCII).trim());
            byte[] IV  = hexToBytes(new String(fileRead(if_iv),  StandardCharsets.US_ASCII).trim());

            System.out.println("[ Java | encrypt ] algorithm : AES-" + (KEY.length * 8) + "/CBC/PKCS7");

            SecretKeySpec secretKey = new SecretKeySpec(KEY, "AES");
            Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
            cipher.init(Cipher.ENCRYPT_MODE, secretKey, new IvParameterSpec(IV));

            byte[] plainText  = fileRead(ifname);
            byte[] cipherText = cipher.doFinal(plainText);
            System.out.println("[ Java | encrypt ] input     : " + plainText.length + " bytes  ->  padded : " + cipherText.length + " bytes");
            fileWrite(ofname, cipherText);
            System.out.println("[ Java | encrypt ] output    : " + ofname);

            byte[] mac = computeHmac(cipherText, KEY);
            fileWrite(ofname + ".sig", mac);
            System.out.println("[ Java | encrypt ] signature : written");

        } catch (java.security.InvalidKeyException e) {
            System.err.println("ABORT: InvalidKeyException — key must be 16, 24, or 32 bytes");
            System.exit(3);
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(4);
        } finally {
            if (KEY != null) Arrays.fill(KEY, (byte) 0);
        }
    }
}
